Implementing reverse routing syntax

burzum opened the pull request:
See these two links

* https://github.com/cakephp/cakephp/issues/10643#issuecomment-301652587
* https://github.com/cakephp/cakephp/issues/10689#issuecomment-304440165

burzum committed:
Implementing reverse routing syntax

stickler-ci committed:
Fixing style errors.

ADmad commented on src/Routing/RouteBuilder.php:
```diff
@@ -703,8 +703,12 @@ public function loadPlugin($name, $file = 'routes.php')
      * @throws \BadMethodCallException
      */
-    public function connect($route, array $defaults = [], array $options = [])
+    public function connect($route, array $defaults = null, array $options = [])
```
Why change the default value?

markstory commented on src/Routing/Router.php:
```diff
@@ -202,14 +202,59 @@ public static function getNamedExpressions()
+                        'prefix' => $matches[2],
+                        'controller' => $matches[3],
+                        'view' => $matches[4]
+                    ];
```
There should probably be an exception thrown for the default case.

markstory commented on tests/TestCase/Routing/RouterTest.php:
```diff
@@ -3422,6 +3422,53 @@ public function testCreateRouteBuilder()
+        ];
+        $this->assertEquals($result, $expected);
+
+        Router::connect('/admin/blog/articles/view', 'Blog.Admin\Articles::view');
```
What about nested prefixes? Those have caused us troubles in other places.

markstory replied:
Or all tokens separated by \ except the last one are the prefix, and the last one is the controller name.

burzum replied:
@markstory like that?

![image](https://user-images.githubusercontent.com/162789/35798348-cfbe787a-0a62-11e8-96db-76e48408215e.png)

markstory replied:
Is `Foo\Bar\Admin` the prefix? Or the plugin name there?

burzum replied:
Prefix? Just specific exactly how you would like to have it implemented? :) Can you just commit a test case for the syntax you want and I'll take care of the rest.

burzum committed:
Doc block fixes and an exception for the new routing syntax

markstory committed:
Move string target parsing into RouteBuilder.

Having it in RouteBuilder makes it accessible to the HTTP method
helpers, as well as connect() on the scoped builder.

Prefix names need to be lower cased, so that has been updated as well.

I've added tests for nested plugins and nested prefixes as well.

markstory commented:
@burzum I've done the changes I think needed to be done and pushed them up.

burzum commented:
@markstory looks good.

markstory commented:
I will merge this once I get the documentation updates done.
